packages/core-data/src/hooks/use-resource-permissions.ts

```diff
@@ -110,14 +114,29 @@ type ResourcePermissionsResolution< IdType > = [
  * @template IdType
  */
 export default function useResourcePermissions< IdType = void >(
-	resource: string,
+	resource: string | EntityResource,
 	id?: IdType
 ): ResourcePermissionsResolution< IdType > {
+	// Serialize `resource` to a string that can be safely used as a React dep.
+	// We can't just pass `resource` as one of the deps, because if it is passed
+	// as an object literal, then it will be a different object on each call even
+	// if the values remain the same.
+	const resourceAsString =
+		typeof resource === 'object' ? addQueryArgs( '', resource ) : resource;
+
 	return useQuerySelect(
 		( resolve ) => {
+			const isEntity = typeof resource === 'object';
+			const hasId = isEntity ? !! resource.id : !! id;
```

Comment:
The hybrid signature won't work. In that case, the `hasId` would be false; the hook would check general permissions for the REST API and return early.

Reply:
Thanks for confirming, but that's even worse. In that case, we need to make sure this behavior is well-documented, and we should probably throw an error if someone calls it with a resource object and a separate ID argument. Because, you know, TS won't be enough sometimes - some projects and developers just don't use it (or abuse it) in their dev environments.

Reply:
Throwing an error in a hook might be a bit risky. Depending on where the hook is called in the tree, the closest error boundary might be the one in the app root, preventing the whole app from loading.

I think logging a `warning` is probably the safest solution here.

Reply:
Logging a `warning` makes sense 👍 I definitely didn't mean throwing a literal error.

Reply:
I added the warning using the `@wordpress/warning` package, which seems to be [the new preferred method](https://github.com/WordPress/gutenberg/pull/63610) for such cases.

cc @gziolo 

## Screenshot
![CleanShot 2024-07-23 at 13 16 49](https://github.com/user-attachments/assets/55391eb1-c5e1-4cd9-992c-b0f73590091a)

Reply:
I’m still processing the usage of `warning` util as I started to have second thoughts of how we can catch these warnings on production. Do we have to? How folks debug the cases where plugins conflict between each other? My reasoning was that if the app continues to work then we don’t need to add noise to the Browser Console when the app runs on production. I’m happy to discuss it further but for block registration it was a clear path forward as it’s mostly targeting the block author when developing the block.

Reply:
In this case, the warning is useful to the developer who's trying to use the API, since it's clear that they're using it in an unexpected way (almost like a front-end counterpart of [`_doing_it_wrong()`](https://developer.wordpress.org/reference/functions/_doing_it_wrong/). To me, it definitely doesn't make sense to surface that in production, it should only be seen in development mode.

Reply:
This warning also has the same target audience - developers. The warnings that aren't actionable by the end user shouldn't be displayed in production environments.

Reply:
I think we reached a consensus here. I appreciate the parallel to `_doint_it_wrong()`. That speaks well to me and maybe we should alias it this way one day to offer more clarity.